File: scripts/llm_orchestrator.py
#!/usr/bin/env python3
"""
Dynamic LLM orchestrator - switches between models by restarting llama-server
on a single shared port. Only one model loaded into VRAM at a time.
"""

import httpx
import subprocess
import signal
import sys
import time
from pathlib import Path

import yaml

from fastapi import FastAPI, Request, HTTPException
from starlette.responses import StreamingResponse

app = FastAPI()

SCRIPT_DIR = Path(__file__).parent.resolve()
PROJECT_DIR = SCRIPT_DIR.parent
YAML_FILE = PROJECT_DIR / "models.yaml"
OS_OVERLAY = PROJECT_DIR / "config" / f"{sys.platform}.yaml"

SHARED_PORT = 18888

CURRENT_MODEL: str | None = None
SERVER_PROCESS: subprocess.Popen | None = None


def load_merged_config() -> dict:
    """Load models.yaml and merge with OS-specific overlay."""
    import json

    base = yaml.safe_load(YAML_FILE.read_text())

    if OS_OVERLAY.exists():
        overlay = yaml.safe_load(OS_OVERLAY.read_text())
        # Merge overlay into base by model name
        for o_model in overlay.get("models", []):
            for b_model in base["models"]:
                if b_model["name"] == o_model["name"]:
                    for k, v in o_model.get("serve", {}).items():
                        b_model.setdefault("serve", {})
                        b_model["serve"][k] = v

    return base


def get_model_config(name: str) -> dict | None:
    """Find model config by name."""
    config = load_merged_config()
    for m in config.get("models", []):
        if m["name"] == name:
            return m
    return None


def find_model_file(local_dir: str) -> Path:
    """Find the .gguf file in local_dir."""
    expanded = Path(local_dir).expanduser()
    files = list(expanded.glob("*.gguf"))
    if not files:
        raise FileNotFoundError(f"No .gguf found in {expanded}")
    return files[0]


def build_server_cmd(model_cfg: dict) -> list[str]:
    """Build llama-server command from model config."""
    serve = model_cfg.get("serve", {})

    local_dir = model_cfg.get("local_dir", "")
    model_file = str(find_model_file(local_dir))

    args = [
        "--model",
        model_file,
        "--host",
        serve.get("host", "127.0.0.1"),
        "--port",
        str(SHARED_PORT),
        "-ngl",
        str(serve.get("n_gpu_layers", 99)),
        "--ctx-size",
        str(serve.get("ctx_size", 8192)),
        "-t",
        str(serve.get("threads", 8)),
        "--temp",
        str(serve.get("temp", 0.7)),
        "--top-p",
        str(serve.get("top_p", 0.95)),
        "--repeat-penalty",
        str(serve.get("repeat_penalty", 1.05)),
    ]

    top_k = serve.get("top_k")
    if top_k is not None:
        args.extend(["--top-k", str(top_k)])

    cache_k = serve.get("cache_type_k")
    if cache_k:
        args.extend(["--cache-type-k", str(cache_k)])

    cache_v = serve.get("cache_type_v")
    if cache_v:
        args.extend(["--cache-type-v", str(cache_v)])

    if serve.get("flash_attn"):
        args.extend(["--flash-attn", "on"])

    if serve.get("jinja"):
        args.append("--jinja")

    batch = serve.get("batch_size")
    if batch is not None:
        args.extend(["--batch-size", str(batch)])

    ubatch = serve.get("ubatch_size")
    if ubatch is not None:
        args.extend(["--ubatch-size", str(ubatch)])

    if serve.get("cont_batching"):
        args.append("--cont-batching")

    mmap = serve.get("mmap")
    if mmap is False:
        args.append("--no-mmap")
    elif mmap is True:
        args.append("--mmap")

    return ["llama-server", *args]


def stop_server():
    """Terminate current llama-server process."""
    global SERVER_PROCESS, CURRENT_MODEL

    if SERVER_PROCESS and SERVER_PROCESS.poll() is None:
        print(f"[{time.strftime('%H:%M:%S')}] Stopping server for '{CURRENT_MODEL}'...")
        SERVER_PROCESS.terminate()
        try:
            SERVER_PROCESS.wait(timeout=15)
        except subprocess.TimeoutExpired:
            SERVER_PROCESS.kill()
            SERVER_PROCESS.wait(timeout=5)
        print(f"[{time.strftime('%H:%M:%S')}] Server stopped, VRAM freed.")

    SERVER_PROCESS = None
    CURRENT_MODEL = None


def wait_for_server():
    """Wait until llama-server is ready (model loaded)."""
    for i in range(120):
        try:
            r = httpx.get(f"http://127.0.0.1:{SHARED_PORT}/health", timeout=2)
            if r.status_code == 200:
                return True
        except Exception:
            pass
        time.sleep(1)

    raise RuntimeError("Server did not become ready within 120s")


def start_server(model_name: str):
    """Start llama-server for the given model."""
    global SERVER_PROCESS, CURRENT_MODEL

    if CURRENT_MODEL == model_name and SERVER_PROCESS and SERVER_PROCESS.poll() is None:
        return  # Already running

    stop_server()

    model_cfg = get_model_config(model_name)
    if not model_cfg:
        raise HTTPException(400, f"Unknown model: {model_name}")

    cmd = build_server_cmd(model_cfg)
    print(f"[{time.strftime('%H:%M:%S')}] Starting server for '{model_name}'...")
    print(f"  CMD: {' '.join(cmd)}")

    SERVER_PROCESS = subprocess.Popen(
        cmd,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    CURRENT_MODEL = model_name

    wait_for_server()
    print(f"[{time.strftime('%H:%M:%S')}] Server for '{model_name}' ready.")


def resolve_model(path: str) -> str:
    """Extract model name from request path."""
    parts = path.split("/")
    for part in parts:
        if part and get_model_config(part):
            return part
    # Default fallback
    default_cfg = load_merged_config()
    for m in default_cfg.get("models", []):
        if m.get("default"):
            return m["name"]
    raise HTTPException(400, "No model specified in path. Use /{model_name}/v1/...")


@app.api_route("/{path:path}", methods=["GET", "POST"])
async def proxy(request: Request, path: str):
    """Route requests to the correct llama-server instance."""
    global CURRENT_MODEL

    target_model = resolve_model(path)

    # Switch model if needed
    if CURRENT_MODEL != target_model:
        start_server(target_model)

    # Strip model name prefix from path
    real_path = path.replace(target_model, "", 1).lstrip("/")

    upstream = f"http://127.0.0.1:{SHARED_PORT}"

    async with httpx.AsyncClient() as client:
        headers = {k: v for k, v in request.headers.items() if k.lower() != "host"}
        resp = await client.request(
            method=request.method,
            url=f"{upstream}/{real_path}",
            headers=headers,
            content=await request.body(),
            timeout=None,
            stream=True,
        )

    return StreamingResponse(
        resp.aiter_bytes(),
        status_code=resp.status_code,
        headers={
            k: v for k, v in resp.headers.items() if k.lower() != "transfer-encoding"
        },
    )


@app.get("/status")
async def status():
    """Return current server status."""
    return {
        "current_model": CURRENT_MODEL,
        "server_running": SERVER_PROCESS is not None and SERVER_PROCESS.poll() is None,
    }


@app.post("/switch/{model_name}")
async def switch(model_name: str):
    """Explicitly switch to a model without sending a request."""
    start_server(model_name)
    return {
        "current_model": CURRENT_MODEL,
        "message": f"Switched to '{model_name}'",
    }


def shutdown(signum, frame):
    print("\nShutting down orchestrator...")
    stop_server()
    sys.exit(0)


signal.signal(signal.SIGINT, shutdown)
signal.signal(signal.SIGTERM, shutdown)

File: scripts/llm_orchestrator.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/exec"
	"os/signal"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"syscall"
	"time"

	"gopkg.in/yaml.v3"
)

type serveConfig struct {
	Host          string  `yaml:"host"`
	Port          int     `yaml:"port"`
	NGPULayers    int     `yaml:"n_gpu_layers"`
	CtxSize       int     `yaml:"ctx_size"`
	Threads       int     `yaml:"threads"`
	Temp          float64 `yaml:"temp"`
	TopP          float64 `yaml:"top_p"`
	TopK          *int    `yaml:"top_k"`
	RepeatPenalty float64 `yaml:"repeat_penalty"`
	CacheTypeK    string  `yaml:"cache_type_k"`
	CacheTypeV    string  `yaml:"cache_type_v"`
	FlashAttn     bool    `yaml:"flash_attn"`
	Jinja         bool    `yaml:"jinja"`
	BatchSize     *int    `yaml:"batch_size"`
	UBatchSize    *int    `yaml:"ubatch_size"`
	ContBatching  bool    `yaml:"cont_batching"`
	Mmap          *bool   `yaml:"mmap"`
}

type modelConfig struct {
	Name     string      `yaml:"name"`
	Repo     string      `yaml:"repo"`
	Include  string      `yaml:"include"`
	LocalDir string      `yaml:"local_dir"`
	Default  bool        `yaml:"default"`
	Serve    serveConfig `yaml:"serve"`
}

type rootConfig struct {
	Models []modelConfig `yaml:"models"`
}

const sharedPort = 18888

var (
	currentModel  string
	serverProcess *exec.Cmd
	serverMu      sync.Mutex
	models        map[string]modelConfig
	defaultModel  string
)

func loadConfig() {
	scriptDir := filepath.Dir(os.Args[0])
	projectDir := filepath.Dir(scriptDir)

	baseFile := filepath.Join(projectDir, "../models.yaml")
	osFile := filepath.Join(projectDir, "config", fmt.Sprintf("%s.yaml", runtime.GOOS))

	models = make(map[string]modelConfig)

	data, err := os.ReadFile(baseFile)
	if err != nil {
		log.Fatalf("Cannot read %s: %v", baseFile, err)
	}

	var cfg rootConfig
	if err := yaml.Unmarshal(data, &cfg); err != nil {
		log.Fatalf("Cannot parse %s: %v", baseFile, err)
	}

	for _, m := range cfg.Models {
		models[m.Name] = m
		if m.Default {
			defaultModel = m.Name
		}
	}

	if overlayData, err := os.ReadFile(osFile); err == nil {
		var overlay rootConfig
		if err := yaml.Unmarshal(overlayData, &overlay); err == nil {
			for _, om := range overlay.Models {
				if bm, ok := models[om.Name]; ok {
					bm.Serve = mergeServe(bm.Serve, om.Serve)
					models[om.Name] = bm
				}
			}
		}
	}
}

func mergeServe(base, overlay serveConfig) serveConfig {
	if overlay.Host != "" {
		base.Host = overlay.Host
	}
	if overlay.NGPULayers > 0 {
		base.NGPULayers = overlay.NGPULayers
	}
	if overlay.CtxSize > 0 {
		base.CtxSize = overlay.CtxSize
	}
	if overlay.Threads > 0 {
		base.Threads = overlay.Threads
	}
	if overlay.Temp > 0 {
		base.Temp = overlay.Temp
	}
	if overlay.TopP > 0 {
		base.TopP = overlay.TopP
	}
	if overlay.TopK != nil {
		base.TopK = overlay.TopK
	}
	if overlay.RepeatPenalty > 0 {
		base.RepeatPenalty = overlay.RepeatPenalty
	}
	if overlay.CacheTypeK != "" {
		base.CacheTypeK = overlay.CacheTypeK
	}
	if overlay.CacheTypeV != "" {
		base.CacheTypeV = overlay.CacheTypeV
	}
	if overlay.FlashAttn {
		base.FlashAttn = true
	}
	if overlay.Jinja {
		base.Jinja = true
	}
	if overlay.BatchSize != nil {
		base.BatchSize = overlay.BatchSize
	}
	if overlay.UBatchSize != nil {
		base.UBatchSize = overlay.UBatchSize
	}
	if overlay.ContBatching {
		base.ContBatching = true
	}
	if overlay.Mmap != nil {
		base.Mmap = overlay.Mmap
	}
	return base
}

func findModelFile(localDir string) (string, error) {
	dir := filepath.Clean(localDir)
	if strings.HasPrefix(dir, "~") {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", err
		}
		dir = filepath.Join(home, dir[1:])
	}

	files, err := filepath.Glob(filepath.Join(dir, "*.gguf"))
	if err != nil || len(files) == 0 {
		return "", fmt.Errorf("no .gguf file found in %s", dir)
	}
	return files[0], nil
}

func buildArgs(cfg modelConfig) []string {
	modelFile, err := findModelFile(cfg.LocalDir)
	if err != nil {
		log.Fatalf("Cannot find model file: %v", err)
	}

	s := cfg.Serve
	args := []string{
		"--model", modelFile,
		"--host", s.Host,
		"--port", fmt.Sprintf("%d", sharedPort),
		"-ngl", fmt.Sprintf("%d", s.NGPULayers),
		"--ctx-size", fmt.Sprintf("%d", s.CtxSize),
		"-t", fmt.Sprintf("%d", s.Threads),
		"--temp", fmt.Sprintf("%.2f", s.Temp),
		"--top-p", fmt.Sprintf("%.2f", s.TopP),
		"--repeat-penalty", fmt.Sprintf("%.2f", s.RepeatPenalty),
	}

	if s.TopK != nil {
		args = append(args, "--top-k", fmt.Sprintf("%d", *s.TopK))
	}
	if s.CacheTypeK != "" {
		args = append(args, "--cache-type-k", s.CacheTypeK)
	}
	if s.CacheTypeV != "" {
		args = append(args, "--cache-type-v", s.CacheTypeV)
	}
	if s.FlashAttn {
		args = append(args, "--flash-attn", "on")
	}
	if s.Jinja {
		args = append(args, "--jinja")
	}
	if s.BatchSize != nil {
		args = append(args, "--batch-size", fmt.Sprintf("%d", *s.BatchSize))
	}
	if s.UBatchSize != nil {
		args = append(args, "--ubatch-size", fmt.Sprintf("%d", *s.UBatchSize))
	}
	if s.ContBatching {
		args = append(args, "--cont-batching")
	}
	if s.Mmap != nil {
		if !*s.Mmap {
			args = append(args, "--no-mmap")
		} else {
			args = append(args, "--mmap")
		}
	}

	return args
}

func stopServer() {
	if serverProcess != nil && serverProcess.Process != nil {
		log.Printf("Stopping server for '%s'...", currentModel)
		serverProcess.Process.Signal(syscall.SIGTERM)
		done := make(chan error, 1)
		go func() { done <- serverProcess.Wait() }()
		select {
		case <-done:
		case <-time.After(10 * time.Second):
			serverProcess.Process.Kill()
			<-done
		}
		log.Println("Server stopped, VRAM freed.")
	}
	serverProcess = nil
	currentModel = ""
}

func waitReady() error {
	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
	defer cancel()

	for {
		select {
		case <-ctx.Done():
			return fmt.Errorf("server did not become ready in time")
		default:
			resp, err := http.Get(fmt.Sprintf("http://127.0.0.1:%d/health", sharedPort))
			if err == nil && resp.StatusCode == 200 {
				resp.Body.Close()
				return nil
			}
			if resp != nil {
				resp.Body.Close()
			}
			time.Sleep(1 * time.Second)
		}
	}
}

func startServer(modelName string) error {
	serverMu.Lock()
	defer serverMu.Unlock()

	if currentModel == modelName && serverProcess != nil && serverProcess.Process != nil && serverProcess.Process.Pid > 0 {
		return nil
	}

	stopServer()

	cfg, ok := models[modelName]
	if !ok {
		return fmt.Errorf("unknown model: %s", modelName)
	}

	args := buildArgs(cfg)
	log.Printf("Starting server for '%s'...", modelName)

	serverProcess = exec.Command("llama-server", args...)
	serverProcess.Stdout = nil
	serverProcess.Stderr = nil
	if err := serverProcess.Start(); err != nil {
		return fmt.Errorf("failed to start llama-server: %w", err)
	}

	currentModel = modelName

	if err := waitReady(); err != nil {
		stopServer()
		return err
	}

	log.Printf("Server for '%s' ready.", modelName)
	return nil
}

func resolveModel(path string) string {
	parts := strings.Split(strings.TrimPrefix(path, "/"), "/")
	for _, part := range parts {
		if part != "" && models[part] != (modelConfig{}) {
			return part
		}
	}
	return defaultModel
}

func stripPrefix(path, model string) string {
	idx := strings.Index(path, "/"+model+"/")
	if idx >= 0 {
		return path[idx+len(model)+2:]
	}
	idx = strings.Index(path, model+"/")
	if idx == 0 {
		return path[len(model)+1:]
	}
	return path
}

type statusResponse struct {
	CurrentModel  string `json:"current_model"`
	ServerRunning bool   `json:"server_running"`
}

type switchResponse struct {
	CurrentModel string `json:"current_model"`
	Message      string `json:"message"`
}

func handler(w http.ResponseWriter, r *http.Request) {
	path := r.URL.Path

	if path == "/status" && r.Method == http.MethodGet {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(statusResponse{
			CurrentModel:  currentModel,
			ServerRunning: serverProcess != nil,
		})
		return
	}

	if strings.HasPrefix(path, "/switch/") && r.Method == http.MethodPost {
		modelName := strings.TrimPrefix(path, "/switch/")
		if err := startServer(modelName); err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(switchResponse{
			CurrentModel: currentModel,
			Message:      fmt.Sprintf("Switched to '%s'", modelName),
		})
		return
	}

	modelName := resolveModel(path)
	realPath := stripPrefix(path, modelName)

	if err := startServer(modelName); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	upstream := fmt.Sprintf("http://127.0.0.1:%d/%s", sharedPort, realPath)

	req, err := http.NewRequestWithContext(r.Context(), r.Method, upstream, r.Body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	for k, vv := range r.Header {
		if k != "Host" {
			req.Header[k] = vv
		}
	}

	client := &http.Client{
		Transport: &http.Transport{
			DisableKeepAlives: false,
		},
	}

	resp, err := client.Do(req)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadGateway)
		return
	}
	defer resp.Body.Close()

	for k, vv := range resp.Header {
		if strings.ToLower(k) != "transfer-encoding" {
			for _, v := range vv {
				w.Header().Add(k, v)
			}
		}
	}

	w.WriteHeader(resp.StatusCode)

	var copyBuf []byte
	bufSize := 32 * 1024
	if r.Method == http.MethodGet {
		bufSize = 64 * 1024
	}

	buf := make([]byte, bufSize)
	for {
		n, err := resp.Body.Read(buf)
		if n > 0 {
			flusher, ok := w.(http.Flusher)
			if !ok {
				copy(copyBuf, buf[:n])
				_, _ = w.Write(buf[:n])
			} else {
				_, _ = w.Write(buf[:n])
				flusher.Flush()
			}
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Printf("Stream error: %v", err)
			break
		}
	}
}

func main() {
	loadConfig()

	log.SetFlags(log.Ltime)

	for _, sig := range []os.Signal{syscall.SIGINT, syscall.SIGTERM} {
		go func(s os.Signal) {
			<-make(chan os.Signal, 1)
			log.Println("\nShutting down orchestrator...")
			stopServer()
			os.Exit(0)
		}(sig)
	}

	// Actually set up signal handling properly
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigCh
		log.Println("\nShutting down orchestrator...")
		stopServer()
		os.Exit(0)
	}()

	addr := fmt.Sprintf(":%d", 8082)
	log.Printf("Orchestrator listening on %s", addr)
	log.Printf("Available models: %v", func() []string {
		names := make([]string, 0, len(models))
		for n := range models {
			names = append(names, n)
		}
		return names
	}())

	if err := http.ListenAndServe(addr, http.HandlerFunc(handler)); err != nil {
		log.Fatalf("Server error: %v", err)
	}
}
